### ISSUE TITLE
Continue spans in TraceHandlerInterceptor

### ISSUE THREAD

**Issue by marcingrzejszczak** [spring-cloud/spring-cloud-sleuth#471]
Currently we're creating a new span. We can consider not creating a new span for this but add all the tags to the previous one. 

Example:

![image](https://cloud.githubusercontent.com/assets/3297437/21102104/e2b764ae-c07c-11e6-8e2d-0794f80fcc7b.png)

We would have 1 span `http/foo` that would have all the tags of `service2-method-in-controller`. 

PROS:
- we're not flooding apps with spans

CONS:
- we're not that explicit in terms of span names

@dsyer @adriancole

**Comment by adriancole** [comment]
thinking this through, flattening the controller lifecycle into annotations could make sense for folks wanting shorter trace depth

If we align the annotation names with the existing tags, it would be easy to follow.

"mvc.controller.start" and finish (annotation/events) would match nicely to the tags "mvc.controller.class" and method

**Comment by marcingrzejszczak** [comment]
@dsyer @adriancole - I would do this for 1.0.12 already since that way it's possible to fix https://github.com/spring-cloud/spring-cloud-sleuth/issues/469 and https://github.com/spring-cloud/spring-cloud-sleuth/issues/427 . We're not breaking API compatibility. The fact is that the spans will look differently (there will be 2 less spans). On the other hand we simplify code A LOT and we can fix two bugs. WDYT?

**Comment by marcingrzejszczak** [comment]
Done via https://github.com/spring-cloud/spring-cloud-sleuth/commit/ee8c73df53478b031e7cd080b328a7b2131b391f

**Comment by nwest1** [comment]
I was experiencing some behavior that I wasn't expecting related to this issue. Overall, it looks like this:
* create root span
* do stuff
* create upstream span with root parent
* inject upstream span
* call service B (instrumented with sleuth)
  * TraceFilter and TraceHandlerInterceptor interact with it
  * Span is closed & reported to zipkin
* upstream span is attempted to be closed, but it's already been overwritten/submitted by sleuth
* root span is closed

I end up with two spans. Sleuth seems to overwrite the service name and instead of creating a new span, it continues the one received in the header. I see that was the intention with this code, but I'm not sure how to let sleuth know to differentiate between an internal parent (continue) or an external parent (new child.) 

If I add a `@NewSpan` to the controller, I see a third span, but sleuth is still continuing the span it receives.

I could be doing this wrong, but here's what I'm seeing:

```
2017-03-22 15:26:42.544 DEBUG [bootstrap,,,] 41749 --- [nio-8080-exec-2] o.s.c.sleuth.instrument.web.TraceFilter  : Received a request to uri [/barcodes/v3/code128s] that should not be sampled [false]
2017-03-22 15:26:42.545 DEBUG [bootstrap,,,] 41749 --- [nio-8080-exec-2] o.s.c.sleuth.instrument.web.TraceFilter  : Found a parent span [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true] in the request
2017-03-22 15:26:42.545 DEBUG [bootstrap,140ea20266c3e4c269ae62cd28754abf,3192d77298d3b01e,true] 41749 --- [nio-8080-exec-2] o.s.c.sleuth.instrument.web.TraceFilter  : Parent span is [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true]
2017-03-22 15:26:42.546 DEBUG [bootstrap,140ea20266c3e4c269ae62cd28754abf,3192d77298d3b01e,true] 41749 --- [nio-8080-exec-2] o.s.c.s.i.web.TraceHandlerInterceptor    : Handling span [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true]
2017-03-22 15:26:42.546 DEBUG [bootstrap,140ea20266c3e4c269ae62cd28754abf,3192d77298d3b01e,true] 41749 --- [nio-8080-exec-2] o.s.c.s.i.web.TraceHandlerInterceptor    : Adding a method tag with value [code128PNG] to a span [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true]
2017-03-22 15:26:42.546 DEBUG [bootstrap,140ea20266c3e4c269ae62cd28754abf,3192d77298d3b01e,true] 41749 --- [nio-8080-exec-2] o.s.c.s.i.web.TraceHandlerInterceptor    : Adding a class tag with value [Code128Controller] to a span [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true]
2017-03-22 15:26:42.547 DEBUG [bootstrap,140ea20266c3e4c269ae62cd28754abf,3192d77298d3b01e,true] 41749 --- [nio-8080-exec-2] o.s.c.s.annotation.DefaultSpanCreator    : For the class [class c.t.esv.barcodes.controller.Code128Controller] method [code128PNG] will name the span [code128-p-n-g]
2017-03-22 15:26:55.703 DEBUG [bootstrap,140ea20266c3e4c269ae62cd28754abf,5d742c7738435f7d,true] 41749 --- [nio-8080-exec-2] .s.c.s.z.ServerPropertiesEndpointLocator : Span will contain serviceName [barcodes-v3]
2017-03-22 15:26:55.706  INFO [bootstrap,140ea20266c3e4c269ae62cd28754abf,3192d77298d3b01e,true] 41749 --- [nio-8080-exec-2] c.t.esv.base.logging.LoggingFilter       : Request Complete
2017-03-22 15:26:55.707 DEBUG [bootstrap,,,] 41749 --- [nio-8080-exec-2] o.s.c.sleuth.instrument.web.TraceFilter  : Trying to send the parent span [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true] to Zipkin
2017-03-22 15:26:55.707 DEBUG [bootstrap,,,] 41749 --- [nio-8080-exec-2] .s.c.s.z.ServerPropertiesEndpointLocator : Span will contain serviceName [barcodes-v3]
2017-03-22 15:26:55.708 DEBUG [bootstrap,,,] 41749 --- [nio-8080-exec-2] o.s.c.sleuth.instrument.web.TraceFilter  : Closing the span [Trace: 140ea20266c3e4c269ae62cd28754abf, Span: 3192d77298d3b01e, Parent: 48cb60f065d5e5b0, exportable:true] since the response was successful
```

And a bit of code:
```
@RestController
@RequestMapping(AppConstants.APP_CONTEXT_ROOT)
class Code128Controller {

    @NewSpan
    @RequestMapping(value = '/code128s', method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    ResponseEntity code128(@RequestBody Code128RequestDTO code128RequestDTO) {
        HttpHeaders headers = HeaderUtils.createCacheHeaders('code128s', 'JSON')
        createImage(code128RequestDTO, headers)
    }
}
```

**Comment by nwest1** [comment]
The more I think about this, the more I'm convinced this is already correct and I'm just thinking about it wrong.

I just had to understand that the parent service is responsible for creating and injecting the child span. I got hung up on what 'closing' a span really means - to each API and to Zipkin.

If I looked more closely at the spring cloud sleuth graphic that shows the flow, that's exactly what is described.

Thanks for the great work!
